File: 169. Majority Element.py
class Solution(object):
    def majorityElement(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if len(nums) == 1:
            return nums[0]
        majority_l = self.majorityElement(nums[:len(nums)/2])
        majority_r = self.majorityElement(nums[len(nums)/2:])
        temp_l = 0
        temp_r = 0
        for i in nums:
            if i == majority_l:
                temp_l += 1
            if i == majority_r:
                temp_r += 1
        if temp_l > len(nums)/2:
            return majority_l
        if temp_r > len(nums)/2:
            return majority_r

File: 169. Majority Element.cpp
class Solution {
public:
    int majorityElement(vector<int>& nums) {
        if (nums.size() == 1)
            return nums[0];
        vector<int> sub_l(nums.begin(), nums.begin() + nums.size()/2);
        vector<int> sub_r(nums.begin() + nums.size()/2, nums.begin() + nums.size());
        int majority_l = majorityElement(sub_l);
        int majority_r = majorityElement(sub_r);
        int temp_l = 0, temp_r = 0;
        for (int i = 0; i < nums.size(); i++)
        {
            if (nums[i] == majority_l)
            {
                temp_l ++;
                if (temp_l > nums.size()/2)
                    return majority_l;
            }
            if (nums[i] == majority_r)
            {
                temp_r ++;
                if (temp_r > nums.size()/2)
                    return majority_r;
            }
        }
        return 0;
    }
};
